File: go/img.go
package main
import (
    "image/color"
    "image/png"
    "os"
    "image"
    "fmt"
)

// image output stuff

func newImg(width, height int) (*image.RGBA, func(int, int, int, int, int)) {
    img := image.NewRGBA(image.Rectangle{image.Point{0, 0}, image.Point{width, height}})
    return img, func(x, y, r, g, b int) {
        // img.Set(x, y, color.RGBA{uint8(rad % 256), uint8(grn % 256), uint8(blu % 256), 0xff})
        img.Set(x, y, color.RGBA{uint8(r % 256), uint8(g % 256), uint8(b % 256), 0xff})
    }
}

func dumpImg(img *image.RGBA) {
    f, err := os.Create(fileName())
    if err != nil {panic(err)}
    png.Encode(f, img) // encode as png
}

func fileName() string {
    for i := 0;; i++ {
        name := fmt.Sprintf("/root/0Git/fracGen/images/image%v.png", i)
        _, err := os.Stat(name)
        if err != nil {return name}
    }
}

// pixel struct

type pixel struct {
    r, g, b float64
    x, y int
}

func (p *pixel) scalecolor(t float64) {
    p.r *= t
    p.g *= t
    p.b *= t
}

func pix(x, y int, r, g, b float64) *pixel {
    p := pixel{}
    p.x, p.y = x, y
    p.r, p.g, p.b = r, g, b
    return &p
}

File: rust/src/img.rs

use image::{ImageBuffer, RgbImage};
use std::path::Path;

pub fn new_img(width: u32, height: u32) -> ImageBuffer<image::Rgb<u8>, std::vec::Vec<u8>> {
    let image: RgbImage = ImageBuffer::new(width, height);
    image
}

// finds what file name is valid
fn file_name() -> String {
    let mut i = 0;
    let mut path: String;
    loop {
        path = format!("../images/image{}.png", i);
        if !Path::new(&path).exists() {return path};
        i += 1;
    }
}

// dumps image with a valid filename
pub fn dump_img(img: ImageBuffer<image::Rgb<u8>, std::vec::Vec<u8>>) {
    img.save(file_name()).unwrap();
}

pub fn dump_img_mut(img: &mut ImageBuffer<image::Rgb<u8>, std::vec::Vec<u8>>) {
    img.save(file_name()).unwrap();
}

// set pixels of image
#[inline(always)]
pub fn set(img: &mut ImageBuffer<image::Rgb<u8>, std::vec::Vec<u8>>, x: u32, y: u32, r: f64, g: f64, b: f64) {
    img.put_pixel(x, y, image::Rgb([r as u8, g as u8, b as u8]))
}

// set pixels of image
#[inline(always)]
pub fn set_u8(img: &mut ImageBuffer<image::Rgb<u8>, std::vec::Vec<u8>>, x: u32, y: u32, r: u8, g: u8, b: u8) {
    img.put_pixel(x, y, image::Rgb([r, g, b]))
}

#[allow(non_camel_case_types)]
pub struct pix {
    pub x: u32,
    pub y: u32,
    pub r: f64,
    pub g: f64,
    pub b: f64,
}

impl pix {

    #[inline(always)]
    pub fn pixel_create(x: u32, y: u32, r: f64, g: f64, b: f64) -> pix {
        pix {
            x,
            y,
            r,
            g,
            b,
        }
    }
}